File: debop4s-data/src/test/scala/com/github/debop4s/data/tests/mapping/associations/ManyToOneTest.scala
package com.github.debop4s.data.tests.mapping.associations

import com.github.debop4s.core.utils.{ToStringHelper, Hashs}
import com.github.debop4s.data.model.HibernateEntity
import com.github.debop4s.data.tests.AbstractJpaTest
import java.lang
import java.util
import javax.persistence._
import org.hibernate.annotations.{LazyCollectionOption, LazyCollection, LazyToOneOption, LazyToOne}
import org.junit.Test

/**
 * ManyToOneTest 
 *
 * @author sunghyouk.bae@gmail.com
 * @since 2014. 3. 3.
 */
@org.springframework.transaction.annotation.Transactional
class ManyToOneTest extends AbstractJpaTest {

    @PersistenceContext val em: EntityManager = null

    @Test
    def unidirectionalManyToOne() {
        val jug = new Jug("JUG Summer Camp")
        val emmanuel = new JugMember("Emmanuel Bernard")
        emmanuel.memberOf = jug

        val jerome = new JugMember("Jerome")
        jerome.memberOf = jug

        em.persist(jug)
        em.persist(emmanuel)
        em.persist(jerome)
        em.flush()
        em.clear()

        val emmanuel2 = em.find(classOf[JugMember], emmanuel.id)
        assert(emmanuel2 != null)
        assert(emmanuel2.id == emmanuel.id)
        assert(emmanuel2.memberOf != null)
        assert(emmanuel2.memberOf == jug)

        em.remove(emmanuel2)
        em.flush()

        val jerome2 = em.find(classOf[JugMember], jerome.id)
        assert(jerome2 != null)
        assert(jerome2.memberOf != null)

        val jug2 = jerome2.memberOf
        em.remove(jerome2)
        em.remove(jug2)
        em.flush()

        assert(em.find(classOf[Jug], jug.id) == null)
    }

}

@Entity
@Access(AccessType.FIELD)
class Beer extends HibernateEntity[lang.Long] {

    @Id
    @GeneratedValue
    var id: lang.Long = _

    def getId = id

    var name: String = _

    var price: lang.Double = _

    @ManyToOne(fetch = FetchType.LAZY, optional = false,
        cascade = Array(CascadeType.MERGE, CascadeType.PERSIST, CascadeType.REFRESH))
    @JoinColumn(nullable = false)
    @LazyToOne(LazyToOneOption.PROXY)
    var brewery: Brewery = _

    override def hashCode(): Int = Hashs.compute(name)

    override protected def buildStringHelper: ToStringHelper =
        super.buildStringHelper
        .add("name", name)
        .add("price", price)
}


@Entity
@Access(AccessType.FIELD)
class Brewery extends HibernateEntity[lang.Long] {

    @Id
    @GeneratedValue
    var id: lang.Long = _

    def getId = id

    var name: String = _

    @OneToMany(cascade = Array(CascadeType.ALL), fetch = FetchType.LAZY)
    @LazyCollection(LazyCollectionOption.EXTRA)
    val beers: util.Set[Beer] = new util.HashSet[Beer]()

    override def hashCode(): Int =
        Hashs.compute(name)

    override protected def buildStringHelper: ToStringHelper =
        super.buildStringHelper
        .add("name", name)
}

@Entity
@Access(AccessType.FIELD)
class Jug extends HibernateEntity[lang.Long] {

    def this(name: String) {
        this()
        this.name = name
    }

    @Id
    @GeneratedValue
    var id: lang.Long = _

    def getId = id

    var name: String = _

    override def hashCode(): Int =
        Hashs.compute(name)

    override protected def buildStringHelper: ToStringHelper =
        super.buildStringHelper
        .add("name", name)
}


@Entity
@Access(AccessType.FIELD)
class JugMember(private[this] val _name: String) extends HibernateEntity[lang.Long] {

    def this() {
        this(null)
    }

    @Id
    @GeneratedValue
    var id: lang.Long = _

    def getId = id

    var name: String = _name

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "memberOf", nullable = false)
    @LazyToOne(LazyToOneOption.PROXY)
    var memberOf: Jug = _

    override def hashCode(): Int =
        Hashs.compute(name)

    override protected def buildStringHelper: ToStringHelper =
        super.buildStringHelper
        .add("name", name)
}

@Entity
@Access(AccessType.FIELD)
class SalesForce extends HibernateEntity[lang.Long] {

    @Id
    @GeneratedValue
    var id: lang.Long = _

    def getId = id

    var corporation: String = _

    @OneToMany(mappedBy = "salesForce", cascade = Array(CascadeType.ALL))
    @LazyCollection(LazyCollectionOption.EXTRA)
    val salesGuys: util.Set[SalesGuy] = new util.HashSet[SalesGuy]

    override def hashCode(): Int =
        Hashs.compute(corporation)

    override protected def buildStringHelper: ToStringHelper =
        super.buildStringHelper
        .add("corporation", corporation)
}

@Entity
@Access(AccessType.FIELD)
class SalesGuy extends HibernateEntity[lang.Long] {

    def this(name: String) {
        this()
        this.name = name
    }

    @Id
    @GeneratedValue
    var id: lang.Long = _

    def getId = id

    var name: String = _

    @ManyToOne(fetch = FetchType.LAZY, cascade = Array(CascadeType.ALL))
    @JoinColumn(name = "salesForceId")
    @LazyToOne(LazyToOneOption.PROXY)
    var salesForce: SalesForce = _

}

File: debop4s-data-orm/src/test/java/debop4s/data/orm/jtests/mapping/associations/manytoone/ManyToOneTest.java
package debop4s.data.orm.jtests.mapping.associations.manytoone;

import debop4s.data.orm.jtests.jpa.JpaTestBase;
import lombok.extern.slf4j.Slf4j;
import org.junit.Test;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;

import static org.fest.assertions.Assertions.assertThat;

@Slf4j
@Transactional
public class ManyToOneTest extends JpaTestBase {

    @PersistenceContext
    EntityManager em;

    @Test
    @Transactional
    public void uniDirectionalManyToOne() {

        Jug jug = new Jug("JUG Summer Camp");

        JugMember emmanuel = new JugMember("Emmanuel Bernard");
        emmanuel.setMemberOf(jug);

        JugMember jerome = new JugMember("Jerome");
        jerome.setMemberOf(jug);

        em.persist(jug);
        em.persist(emmanuel);
        em.persist(jerome);
        em.flush();
        em.clear();

        emmanuel = em.find(JugMember.class, emmanuel.getId());
        assertThat(emmanuel).isNotNull();

        jug = emmanuel.getMemberOf();
        assertThat(jug).isNotNull();

        em.remove(emmanuel);

        jerome = em.find(JugMember.class, jerome.getId());

        assertThat(jerome).isNotNull();

        em.remove(jerome);
        em.remove(jug);
        em.flush();

        assertThat(em.find(JugMember.class, jerome.getId())).isNull();
    }

    @Test
    @Transactional
    public void testBidirectionalManyToOneRegular() throws Exception {

        SalesForce force = new SalesForce("Red Hat");
        em.persist(force);

        SalesGuy eric = new SalesGuy();
        eric.setName("Eric");
        eric.setSalesForce(force);
        force.getSalesGuys().add(eric);
        em.persist(eric);

        SalesGuy simon = new SalesGuy();
        simon.setName("Simon");
        simon.setSalesForce(force);
        force.getSalesGuys().add(simon);
        em.persist(simon);

        em.flush();
        em.clear();

        force = em.find(SalesForce.class, force.getId());
        assertThat(force.getSalesGuys()).isNotNull();
        assertThat(force.getSalesGuys().size()).isEqualTo(2);
        for (SalesGuy guy : force.getSalesGuys()) {
            assertThat(guy.getId()).isNotNull();
        }

        simon = em.find(SalesGuy.class, simon.getId());
        assertThat(simon).isNotNull();

        // Cascade 때문에
        force.getSalesGuys().remove(simon);
        em.remove(simon);
        em.persist(force);
        em.flush();
        em.clear();

        force = em.find(SalesForce.class, force.getId());
        assertThat(force.getSalesGuys()).isNotNull();
        assertThat(force.getSalesGuys().size()).isEqualTo(1);
        assertThat(force.getSalesGuys().contains(simon)).isFalse();
        assertThat(force.getSalesGuys().contains(eric)).isTrue();

        em.remove(force);
        em.flush();

        assertThat(em.find(SalesGuy.class, force.getSalesGuys().iterator().next().getId())).isNull();
    }

    @Test
    @Transactional
    public void testBiDirManyToOneInsertUpdateFalse() throws Exception {

        Brewery hoeBrewery = new Brewery();
        Beer hoegaarden = new Beer();
        hoeBrewery.getBeers().add(hoegaarden);
        hoegaarden.setBrewery(hoeBrewery);
        em.persist(hoeBrewery);
        em.flush();
        em.clear();

        hoegaarden = em.find(Beer.class, hoegaarden.getId());
        assertThat(hoegaarden).isNotNull();
        assertThat(hoegaarden.getBrewery()).isNotNull();
        assertThat(hoegaarden.getBrewery().getId()).isNotNull();
        assertThat(hoegaarden.getBrewery().getBeers()).hasSize(1).containsOnly(hoegaarden);

        Beer citron = new Beer();
        hoeBrewery = hoegaarden.getBrewery();
        hoeBrewery.getBeers().remove(hoegaarden);
        hoeBrewery.getBeers().add(citron);
        citron.setBrewery(hoeBrewery);
        em.remove(hoegaarden);
        em.flush();
        em.clear();

        citron = em.find(Beer.class, citron.getId());
        assertThat(citron).isNotNull();
        assertThat(citron.getBrewery().getBeers()).hasSize(1).containsOnly(citron);

        hoeBrewery = citron.getBrewery();
        citron.setBrewery(null);
        hoeBrewery.getBeers().clear();

        em.remove(citron);
        em.remove(hoeBrewery);
        em.flush();

        assertThat(em.find(Beer.class, citron.getId())).isNull();
    }
}
